## Домашнее задание №1 по блоку CV

1. Возьмите https://www.kaggle.com/ajayrana/hymenoptera-data

In [1]:
import cv2
import numpy as np
import tensorflow as tf
import tensorflow.keras as keras
from random import shuffle
from warnings import filterwarnings 
filterwarnings('ignore')

2. Реализуйте сверточную нейронную сеть с использованием tf.keras или keras. Используйте сверточные, пуллинговые и полносвязанные слои. Обучите на train выборке в течении 10 эпох, оцените качество на отложенной выборке

In [2]:
train_generator = keras.preprocessing.image.ImageDataGenerator()
test_generator = keras.preprocessing.image.ImageDataGenerator()

In [3]:
train = train_generator.flow_from_directory('hymenoptera_data/train', 
                                            target_size=(224, 224), 
                                            batch_size=32, 
                                            class_mode='binary')

test = test_generator.flow_from_directory('hymenoptera_data/val', 
                                          target_size=(224, 224), 
                                          batch_size=32, 
                                          class_mode='binary')

Found 244 images belonging to 2 classes.
Found 153 images belonging to 2 classes.


In [4]:
i = keras.layers.Input(shape=(224,224,3))
o = keras.layers.Conv2D(16, (5,5), strides=(3,3))(i) 
o = keras.layers.MaxPool2D((2,2))(o)
o = keras.layers.ReLU()(o)
o = keras.layers.Conv2D(32, (3,3))(o)
o = keras.layers.MaxPool2D((2,2))(o)
o = keras.layers.ReLU()(o)
o = keras.layers.Conv2D(64, (3,3), strides=(1,1))(o)
o = keras.layers.MaxPool2D((2,2))(o)
o = keras.layers.ReLU()(o)
o = keras.layers.Flatten()(o)
o = keras.layers.Dense(1, activation='sigmoid')(o)

In [5]:
model = keras.Model(inputs=i, outputs=o)
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
model.summary()

Model: "model"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         [(None, 224, 224, 3)]     0         
_________________________________________________________________
conv2d (Conv2D)              (None, 74, 74, 16)        1216      
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 37, 37, 16)        0         
_________________________________________________________________
re_lu (ReLU)                 (None, 37, 37, 16)        0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 35, 35, 32)        4640      
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 17, 17, 32)        0         
_________________________________________________________________
re_lu_1 (ReLU)               (None, 17, 17, 32)        0     

In [6]:
model.fit(train, epochs=10, validation_data=test)

Epoch 1/10
8/8 [==============================] - 3s 308ms/step - loss: 15.6965 - accuracy: 0.5013 - val_loss: 2.4790 - val_accuracy: 0.5294
Epoch 2/10
8/8 [==============================] - 2s 259ms/step - loss: 1.4748 - accuracy: 0.5576 - val_loss: 1.3021 - val_accuracy: 0.5556
Epoch 3/10
8/8 [==============================] - 2s 254ms/step - loss: 0.9625 - accuracy: 0.6328 - val_loss: 0.9352 - val_accuracy: 0.5817
Epoch 4/10
8/8 [==============================] - 2s 255ms/step - loss: 0.7159 - accuracy: 0.6503 - val_loss: 0.8599 - val_accuracy: 0.5686
Epoch 5/10
8/8 [==============================] - 2s 252ms/step - loss: 0.5029 - accuracy: 0.7625 - val_loss: 0.7623 - val_accuracy: 0.6209
Epoch 6/10
8/8 [==============================] - 2s 254ms/step - loss: 0.3818 - accuracy: 0.8638 - val_loss: 0.7431 - val_accuracy: 0.6405
Epoch 7/10
8/8 [==============================] - 2s 247ms/step - loss: 0.3252 - accuracy: 0.8976 - val_loss: 0.7365 - val_accuracy: 0.6471
Epoch 8/10
8/8 [===

Качество на валидационной выборке так себе val_loss: 0.7752 - val_accuracy: 0.627

3. Добавьте в предыдущую архитектуру слои BatchNorm. Обучите на train выборке в течении 10 эпох, оцените качество на отложенной выборке

In [7]:
i = keras.layers.Input(shape=(224,224,3))
o = keras.layers.BatchNormalization()(i)
o = keras.layers.Conv2D(16, (5,5), strides=(3,3))(o)
o = keras.layers.MaxPool2D((2,2))(o)
o = keras.layers.ReLU()(o)
o = keras.layers.BatchNormalization()(o)
o = keras.layers.Conv2D(32, (3,3))(o)
o = keras.layers.MaxPool2D((2,2))(o)
o = keras.layers.ReLU()(o)
o = keras.layers.BatchNormalization()(o)
o = keras.layers.Flatten()(o)
o = keras.layers.Dense(1, activation='sigmoid')(o)

In [8]:
model = keras.Model(inputs=i, outputs=o)
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
model.summary()

Model: "model_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_2 (InputLayer)         [(None, 224, 224, 3)]     0         
_________________________________________________________________
batch_normalization (BatchNo (None, 224, 224, 3)       12        
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 74, 74, 16)        1216      
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 37, 37, 16)        0         
_________________________________________________________________
re_lu_3 (ReLU)               (None, 37, 37, 16)        0         
_________________________________________________________________
batch_normalization_1 (Batch (None, 37, 37, 16)        64        
_________________________________________________________________
conv2d_4 (Conv2D)            (None, 35, 35, 32)        4640

In [9]:
model.fit(train, epochs=10, validation_data=test)

Epoch 1/10
8/8 [==============================] - 4s 475ms/step - loss: 1.0274 - accuracy: 0.4666 - val_loss: 0.9400 - val_accuracy: 0.5425
Epoch 2/10
8/8 [==============================] - 4s 439ms/step - loss: 0.5388 - accuracy: 0.7486 - val_loss: 1.0434 - val_accuracy: 0.5752
Epoch 3/10
8/8 [==============================] - 4s 440ms/step - loss: 0.2623 - accuracy: 0.9015 - val_loss: 1.9621 - val_accuracy: 0.5425
Epoch 4/10
8/8 [==============================] - 3s 431ms/step - loss: 0.1581 - accuracy: 0.9563 - val_loss: 0.9670 - val_accuracy: 0.5882
Epoch 5/10
8/8 [==============================] - 3s 428ms/step - loss: 0.1007 - accuracy: 0.9900 - val_loss: 0.9037 - val_accuracy: 0.5556
Epoch 6/10
8/8 [==============================] - 4s 438ms/step - loss: 0.0652 - accuracy: 0.9871 - val_loss: 0.7253 - val_accuracy: 0.6144
Epoch 7/10
8/8 [==============================] - 4s 436ms/step - loss: 0.0496 - accuracy: 1.0000 - val_loss: 0.7274 - val_accuracy: 0.6209
Epoch 8/10
8/8 [====

После слоя нормализации стало совсем чуть-чуть получше val_loss: 0.6240 - val_accuracy: 0.699

4. Создайте модель ResNet 50, инициализированную случайными весами, и обучите ее на train выборке в течении 10 эпох, оцените качество на отложенной выборке

In [10]:
train_generator = keras.preprocessing.image.ImageDataGenerator()
test_generator  = keras.preprocessing.image.ImageDataGenerator()

train = train_generator.flow_from_directory('hymenoptera_data/train', 
                                            target_size= (224, 224), 
                                            batch_size = 32, 
                                            class_mode ='binary')

test  = test_generator.flow_from_directory('hymenoptera_data/val', 
                                           target_size = (224, 224), 
                                           batch_size  = 32, 
                                           class_mode  ='binary')

Found 244 images belonging to 2 classes.
Found 153 images belonging to 2 classes.


In [11]:
resnet50 = keras.applications.ResNet50(include_top=False, 
                                       weights=None, 
                                       input_tensor=None, 
                                       input_shape=None, 
                                       pooling=None, 
                                       classes=1000 
                                       )

In [12]:
i = keras.layers.Input((224, 224, 3), dtype = tf.uint8)
o = tf.cast(i, tf.float32)
o = resnet50(o)
o = keras.layers.Flatten()(o)
o = keras.layers.Dense(1, activation='sigmoid')(o)

In [13]:
model = keras.Model(inputs=i, outputs=o)
model.compile(optimizer='Adagrad', loss='binary_crossentropy', metrics=['accuracy']) 
model.summary()

Model: "model_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_4 (InputLayer)         [(None, 224, 224, 3)]     0         
_________________________________________________________________
tf.cast (TFOpLambda)         (None, 224, 224, 3)       0         
_________________________________________________________________
resnet50 (Functional)        (None, None, None, 2048)  23587712  
_________________________________________________________________
flatten_2 (Flatten)          (None, 100352)            0         
_________________________________________________________________
dense_2 (Dense)              (None, 1)                 100353    
Total params: 23,688,065
Trainable params: 23,634,945
Non-trainable params: 53,120
_________________________________________________________________


In [14]:
model.fit(train, epochs=10, validation_data=test)

Epoch 1/10
8/8 [==============================] - 78s 9s/step - loss: 8.5511 - accuracy: 0.4857 - val_loss: 1.0090 - val_accuracy: 0.6275
Epoch 2/10
8/8 [==============================] - 75s 9s/step - loss: 3.5373 - accuracy: 0.5688 - val_loss: 1.1012 - val_accuracy: 0.4641
Epoch 3/10
8/8 [==============================] - 74s 9s/step - loss: 2.4835 - accuracy: 0.6062 - val_loss: 4.0387 - val_accuracy: 0.5425
Epoch 4/10
8/8 [==============================] - 75s 10s/step - loss: 1.8758 - accuracy: 0.5506 - val_loss: 1.5878 - val_accuracy: 0.5882
Epoch 5/10
8/8 [==============================] - 74s 9s/step - loss: 0.7407 - accuracy: 0.6804 - val_loss: 1.4132 - val_accuracy: 0.5817
Epoch 6/10
8/8 [==============================] - 76s 9s/step - loss: 0.6006 - accuracy: 0.7223 - val_loss: 0.7584 - val_accuracy: 0.6340
Epoch 7/10
8/8 [==============================] - 75s 9s/step - loss: 0.6259 - accuracy: 0.6761 - val_loss: 0.7126 - val_accuracy: 0.5229
Epoch 8/10
8/8 [=================

Очень низкое качество модели val_loss: 0.7749 - val_accuracy: 0.4641

5. Создайте модель ResNet 50, инициализированную весами ImageNet, и обучите ее на train выборке в течении 10 эпох, оцените качество на отложенной выборке

In [15]:
resnet50 = keras.applications.ResNet50(include_top=False, 
                                       weights='imagenet', 
                                       input_tensor=None, 
                                       input_shape=None, 
                                       pooling=None, 
                                       classes=1000
                                       )

In [16]:
i = keras.layers.Input((224, 224, 3), dtype = tf.uint8)
o = tf.cast(i, tf.float32)
o = resnet50(o)
o = keras.layers.Flatten()(o)
o = keras.layers.Dense(1, activation='sigmoid')(o)

In [17]:
resnet50.trainable = False

In [18]:
model = keras.Model(inputs=i, outputs=o)
model.compile(optimizer='Adagrad', loss='binary_crossentropy', metrics=['accuracy'])
model.summary()

Model: "model_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_6 (InputLayer)         [(None, 224, 224, 3)]     0         
_________________________________________________________________
tf.cast_1 (TFOpLambda)       (None, 224, 224, 3)       0         
_________________________________________________________________
resnet50 (Functional)        (None, None, None, 2048)  23587712  
_________________________________________________________________
flatten_3 (Flatten)          (None, 100352)            0         
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 100353    
Total params: 23,688,065
Trainable params: 100,353
Non-trainable params: 23,587,712
_________________________________________________________________


In [19]:
model.fit(train, epochs=10, validation_data=test)

Epoch 1/10
8/8 [==============================] - 15s 2s/step - loss: 3.2816 - accuracy: 0.5604 - val_loss: 0.5847 - val_accuracy: 0.8301
Epoch 2/10
8/8 [==============================] - 13s 2s/step - loss: 0.1345 - accuracy: 0.9514 - val_loss: 0.4881 - val_accuracy: 0.8562
Epoch 3/10
8/8 [==============================] - 13s 2s/step - loss: 0.0259 - accuracy: 0.9871 - val_loss: 0.5563 - val_accuracy: 0.8301
Epoch 4/10
8/8 [==============================] - 13s 2s/step - loss: 0.0091 - accuracy: 1.0000 - val_loss: 0.4956 - val_accuracy: 0.8562
Epoch 5/10
8/8 [==============================] - 13s 2s/step - loss: 0.0049 - accuracy: 1.0000 - val_loss: 0.5030 - val_accuracy: 0.8562
Epoch 6/10
8/8 [==============================] - 13s 2s/step - loss: 0.0046 - accuracy: 1.0000 - val_loss: 0.4962 - val_accuracy: 0.8562
Epoch 7/10
8/8 [==============================] - 13s 2s/step - loss: 0.0044 - accuracy: 1.0000 - val_loss: 0.4938 - val_accuracy: 0.8562
Epoch 8/10
8/8 [==================

С весами ImageNet получалась самая удачная модель val_loss: 0.4971 - val_accuracy: 0.8562